lib/HLSL/DxilPrecisePropagatePass.cpp

```diff
@@ -51,73 +54,169 @@ class DxilPrecisePropagatePass : public ModulePass {
+          continue;
+        PropagatePreciseAttributeOnPointerUsers(GEP, typeSys, Context, processedSet);
+      } else {
+        // Recurse GEP users
+        PropagatePreciseAttributeThroughGEPs(
+            GEP, typeSys, Context,
+            ArrayRef<Value*>(idxList.data() + i, idxList.end()),
+            processedGEPs, processedSet);
+      }
+    }
+  }
+}
+
 static void PropagatePreciseAttributeOnPointer(
     Value *Ptr, DxilTypeSystem &typeSys, LLVMContext &Context,
-    std::unordered_set<Instruction *> &processedSet) {
+    ValueSet &processedSet) {
+
+  PropagatePreciseAttributeOnPointerUsers(Ptr, typeSys, Context, processedSet);
+
+  if (GEPOperator *GEP = dyn_cast<GEPOperator>(Ptr)) {
+    // Get root Ptr, gather index list, and mark matching stores
+    SmallVector<Value*, 8> idxList;
+    SmallVector<GEPOperator*, 4> GEPs;
+    GEPs.emplace_back(GEP);
+    while (GEP = dyn_cast<GEPOperator>(Ptr = GEP->getPointerOperand()))
+      GEPs.emplace_back(GEP);
+    while (!GEPs.empty()) {
+      GEP = GEPs.back();
+      GEPs.pop_back();
+      auto idx = GEP->idx_begin();
+      idx++;
+      while (idx != GEP->idx_end())
+        idxList.emplace_back(*(idx++));
+    }
+    ValueSet processedGEPs;
+    PropagatePreciseAttributeThroughGEPs(
+        Ptr, typeSys, Context,
+        idxList, processedGEPs, processedSet);
+  }
+}
+
+static void PropagatePreciseAttributeOnPointerUsers(
+    Value *Ptr, DxilTypeSystem &typeSys, LLVMContext &Context,
+    ValueSet &processedSet) {
   // Find all store and propagate on the val operand of store.
   // For CallInst, if Ptr is used as out parameter, mark it.
   for (User *U : Ptr->users()) {
-    Instruction *user = cast<Instruction>(U);
-    if (StoreInst *stInst = dyn_cast<StoreInst>(user)) {
+    if (StoreInst *stInst = dyn_cast<StoreInst>(U)) {
       Value *val = stInst->getValueOperand();
       PropagatePreciseAttributeOnOperand(val, typeSys, Context, processedSet);
-    } else if (CallInst *CI = dyn_cast<CallInst>(user)) {
+    } else if (CallInst *CI = dyn_cast<CallInst>(U)) {
       bool bReadOnly = true;
 
       Function *F = CI->getCalledFunction();
+
+      // skip starting points (dx.attribute.precise calls)
+      if (HLModule::HasPreciseAttribute(F))
+        return;
+
       const DxilFunctionAnnotation *funcAnnotation =
           typeSys.GetFunctionAnnotation(F);
-      for (unsigned i = 0; i < CI->getNumArgOperands(); ++i) {
-        if (Ptr != CI->getArgOperand(i))
-          continue;
 
-        const DxilParameterAnnotation &paramAnnotation =
-            funcAnnotation->GetParameterAnnotation(i);
-        // OutputPatch and OutputStream will be checked after scalar repl.
-        // Here only check out/inout
-        if (paramAnnotation.GetParamInputQual() == DxilParamInputQual::Out ||
-            paramAnnotation.GetParamInputQual() == DxilParamInputQual::Inout) {
-          bReadOnly = false;
-          break;
+      if (funcAnnotation) {
```

Comment:
Would be great to have a test for this. I was actually not clear on what the expected behavior is here.

Are these two shaders equivalent in the precise behavior?
```
precise float4 MakePrecise(precise float4 v) { return v; }

float4 VS(out pos : SV_Position)
{
   ...
   return MakePrecise(...)
}
```

vs.

```
float4 VS(precise out pos : SV_Position)
{
   ...
   return ...;
}

```

Reply:
The MakePrecise function defined here does not work (at least not in all cases).  However, it can be modified to work by declaring a precise local variable and returning that instead.  I'll update the test to use this, since marking the output field is tested elsewhere.  It would be good if the original MakePrecise definition worked, so we should probably file an issue to follow up on that.